src/v/raft/seq_stm.cc

```diff
@@ -0,0 +1,348 @@
+                    "({}) should have matched it or bypassed",
+                    target_offset,
+                    _insync_offset);
+                  return do_make_snapshot();
+              });
+        });
+    });
+}
+
+ss::future<> seq_stm::do_make_snapshot() {
+    vlog(raftlog.trace, "saving seq_stm snapshot");
+    snapshot data;
+    data.offset = _insync_offset;
+    for (auto& entry : _seq_table) {
+        data.entries.push_back(entry.second);
+    }
+
+    iobuf v_buf;
+    reflection::adl<snapshot>{}.to(v_buf, data);
+    return persist_snapshot(std::move(v_buf))
+      .then([this, snapshot_offset = data.offset]() {
+          if (snapshot_offset >= _last_snapshot_offset) {
+              _last_snapshot_offset = snapshot_offset;
+          }
+          return _snapshot_mgr.remove_partial_snapshots();
+      });
+}
+
+ss::future<> seq_stm::start() {
+    _oldest_session = model::timestamp::now();
+    return _snapshot_mgr.open_snapshot().then(
+      [this](std::optional<storage::snapshot_reader> reader) {
+          auto f = ss::now();
+          if (reader) {
+              f = ss::do_with(
+                std::move(*reader), [this](storage::snapshot_reader& reader) {
+                    return hydrate_snapshot(reader).finally([this, &reader] {
+                        auto offset = std::max(
+                          _insync_offset, _c->start_offset());
+                        if (offset >= model::offset(0)) {
+                            set_next(offset);
+                        }
+                        _resolved_when_snapshot_hydrated.set_value();
+                        return reader.close();
+                    });
+                });
+          } else {
+              auto offset = _c->start_offset();
+              if (offset >= model::offset(0)) {
+                  set_next(offset);
+              }
+              _resolved_when_snapshot_hydrated.set_value();
+          }
+
+          return f.then([this]() { return state_machine::start(); })
+            .then([this]() {
+                auto offset = _c->meta().commit_index;
+                if (offset >= model::offset(0)) {
+                    (void)ss::with_gate(_gate, [this, offset] {
+                        // saving a snapshot after catchup with the tip of the
+                        // log
+                        return ensure_snapshot_exists(offset);
+                    });
+                }
+            });
+      });
+}
+
+ss::future<> seq_stm::catchup() {
+    if (_insync_term != _c->term()) {
+        if (!_is_catching_up) {
+            _is_catching_up = true;
+            auto meta = _c->meta();
+            return catchup(meta.prev_log_term, meta.prev_log_index)
+              .then([this]() { _is_catching_up = false; });
+        }
+    }
+    return ss::now();
+}
+
+static int64_t time_since_epoch() {
+    return std::chrono::duration_cast<std::chrono::milliseconds>(
```

Comment:
this is already implemente in model/timestamp.h

you can use `model::timestamp::now()` to get an epoch timestamp and it is a strong type. 

migth be worth using the strong type